SubEthaEdit-Mac/Source/TabbedDocument.m

```diff
@@ -0,0 +1,145 @@
+           then:nil];
+}
+
+- (void)warn:(NSString *)message
+     details:(NSString *)details
+     buttons:(NSArray *)buttons
+        then:(AlertConsequence)then {
+    [self alert:message
+          style:NSAlertStyleWarning
+        details:details
+        buttons:buttons
+           then:then];
+}
+
+- (void)presentAlert:(NSAlert *)alert completionHandler:(void (^)(NSModalResponse returnCode))completionHandler {
+    NSArray *orderedWindows = NSApp.orderedWindows;
+    NSSet *candidateWindows = [NSSet setWithArray:[self.windowControllers valueForKey:@"window"]];
+
+    NSUInteger index = [orderedWindows indexOfObjectPassingTest:
+                        ^(NSWindow *window, NSUInteger idx, BOOL *_stop) {
+                            return [candidateWindows containsObject:window];
+                        }];
+    NSWindow *window = orderedWindows[index];
+
+    [window makeKeyAndOrderFront:self];
+    [alert beginSheetModalForWindow:window completionHandler:completionHandler];
+}
+
+@end
+
+
+@implementation TabbedDocument (SheetSynchronization)
+
+// We collect all the notification names and selectors we care about in this method
+// to avoid potential bugs of forgetting to remove an observer we've added.
+- (NSDictionary *)windowNotifications {
+    return @{
+             NSWindowWillBeginSheetNotification:
+                 NSStringFromSelector(@selector(windowWillBeginSheet:)),
+             NSWindowDidEndSheetNotification:
+                 NSStringFromSelector(@selector(windowDidEndSheet:))
+             };
+}
+
+- (void)addWindowController:(NSWindowController *)windowController {
+    [super addWindowController:windowController];
+
+    NSNotificationCenter *defaultCenter = NSNotificationCenter.defaultCenter;
+    NSDictionary *notifications = self.windowNotifications;
+    for (NSString *name in notifications)
+        [defaultCenter addObserver:self
+                          selector:NSSelectorFromString(notifications[name])
+                              name:name
+                            object:windowController.window];
+}
+
+- (void)removeWindowController:(NSWindowController *)windowController {
+    [super addWindowController:windowController];
+
+    NSNotificationCenter *defaultCenter = NSNotificationCenter.defaultCenter;
+    for (NSString *name in self.windowNotifications)
+        [defaultCenter removeObserver:self
+                             name:name
+                            object:windowController.window];
+}
+
+
+// We actually would like window*Did*BeginSheet:, but unfortunately that notification
+// doesn't exist, so at this point window.attachedSheet is still nil. For this reason
+// the willChangeValueForKey: is correct, but we have to delay the didChangeValueForKey:.
+- (void)windowWillBeginSheet:(NSNotification *)notification {
+    [self willChangeValueForKey:@"hasAlerts"];
+
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
```

Comment:
please use the `[NSOperationQueue TCM_performBlockOnMainQueue:afterDelay:]` for this so these kind of calls have a common callsite to search for.

Reply:
Fixed: https://github.com/subethaedit/SubEthaEdit/pull/128/commits/dd94a8c7961b1614ad223c3f8554e91b7992f17e